### PR TITLE
Add precision to support magnitudes less than 1

### PR DESCRIPTION
**Pull request by cyrossignol**
This adds support for superblock storage of magnitude values as small as 0.01 by increasing the precision of lesser magnitude values to better represent smaller crunchers. Currently, a CPID with a magnitude of 0.5 or less will fall out of the superblock because these magnitudes round-down to zero (greater than 0.5 round-up to 1). After the removal of the team requirement, an influx of new participants may affect the magnitude distribution in a way that makes it difficult for participants with less computing power to earn research rewards by preventing those will lower magnitude from appearing in superblocks. 

We discussed several strategies for storing magnitudes with more precision to mitigate this. I kept my initial implementation with these magnitude precision tiers:

 - Less than 1 exhibit two fractional places 
 - 1 or greater and less than 10 exhibit one fractional place
 - 10 or greater are whole numbers only

This balances the fidelity of magnitude values with storage overhead. Magnitude values are currently encoded as 16-bit integers (two bytes), and the new superblock format can compress magnitudes smaller than 253 into one byte. These precision tiers allow us to continue to use this compression for magnitudes up to 253 even with the increased precision on the low end. The majority of magnitudes in the network fall below this threshold. 

More importantly, these tiers allow for more deterministic superblock sizes. With a strategy that uses both bytes to embed more precision, we can support two fractional places for magnitudes up to about 327. However, this severely limits the range of values that fit into one byte (it caps at around 1.27 magnitude). A powerful cruncher that disappears from a superblock can cause a large fluctuation in superblock size as the magnitudes for many smaller crunchers jump up from one to two bytes. If the network reaches capacity, that sudden jump in size will cause thousands of small CPIDs to fall out of a superblock because of the size limit. For better determinism, we can also serialize magnitudes as two bytes always, but using compression allows for over 2800 more participants before reaching the limit and reduces the rate of growth of the blockchain.

To provide a way to identify serialized magnitudes by precision tier, we talked about using a specialized encoding of the integers that designate some of the bits as flags. Instead, I chose to redesign the magnitude collection in the superblock data structure by splitting it into segments--one for each magnitude size category. Each segment contains the compact representations of the magnitudes in that category. This adds a small cost to scale the magnitudes in a superblock for use by the application, but, in return, bulk superblock data serializes/deserializes much more efficiently without the overhead of the special encoding. 

We have room to adjust the thresholds for the proposed precision tiers to something like _0 &rarr; 2.5 &rarr; 25 &rarr; 32767_ to increase the range of values with higher precision. As recommended by @jamescowens, I started with the powers of 10 for simplicity. I'm interested to hear thoughts about these tiers.

### CODE REVIEW

**Comment by jamescowens** [comment]
Haven't looked at the code yet. Conceptually this is a wonderful way to implement this... essentially normalizes the category ("shift" bits) rather than repeating them over and over in each value stored.

As you know, I am a big advocate of powers of 10 separation between the precision tiers, because that is how normal "significant digits" work in the scientific word, and they are easy to remember.

Note the 32-bit builds are failing, so without even looking at the code, there may be an architecture dependent length issue we have to resolve.

**Comment by jamescowens** [comment]
From travis...

In file included from ./neuralnet/cpid.h:3:0,
                 from ./neuralnet/superblock.h:3,
                 from test/neuralnet/superblock_tests.cpp:2:
test/neuralnet/superblock_tests.cpp: In member function ‘void Superblock::it_serializes_to_a_stream::test_method()’:
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:863:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Small magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:866:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Medium magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:869:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Large magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:874:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(meta.project_count)              // Project stats
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
test/neuralnet/superblock_tests.cpp: In member function ‘void Superblock::it_deserializes_from_a_stream::test_method()’:
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:907:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Small magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:910:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Medium magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:913:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Large magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:918:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(meta.project_count)              // Project stats
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
test/neuralnet/superblock_tests.cpp: In member function ‘void Superblock::it_serializes_to_a_stream_for_fallback_convergences::test_method()’:
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:982:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Small magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:985:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Medium magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:988:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Large magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:993:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(meta.project_count)              // Project stats
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
test/neuralnet/superblock_tests.cpp: In member function ‘void Superblock::it_deserializes_from_a_stream_for_fallback_convergence::test_method()’:
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1032:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Small magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1035:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Medium magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1038:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Large magnitudes
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1043:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(meta.project_count)              // Project stats
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
test/neuralnet/superblock_tests.cpp: In member function ‘void Superblock__CpidIndex::it_serializes_to_a_stream::test_method()’:
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1422:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{0})                       // Small CPIDs size
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1423:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{0})                       // Medium CPIDs size
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1424:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Large CPIDs size
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
test/neuralnet/superblock_tests.cpp: In member function ‘void Superblock__CpidIndex::it_deserializes_from_a_stream::test_method()’:
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1452:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{0})                       // Small CPIDs size
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1453:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{0})                       // Medium CPIDs size
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
./serialize.h:428:42: error: cannot bind non-const lvalue reference of type ‘uint64_t& {aka long long unsigned int&}’ to an rvalue of type ‘uint64_t {aka long long unsigned int}’
 #define COMPACTSIZE(obj) CCompactSize(REF(obj))
                                       ~~~^~~~~
test/neuralnet/superblock_tests.cpp:1454:12: note: in expansion of macro ‘COMPACTSIZE’
         << COMPACTSIZE(size_t{1})                       // Large CPIDs size
            ^
./serialize.h:489:14: note:   initializing argument 1 of ‘CCompactSize::CCompactSize(uint64_t&)’
     explicit CCompactSize(uint64_t& nIn) : n(nIn) { }
              ^~~~~~~~~~~~
Makefile:7784: recipe for target 'test/neuralnet/test_test_gridcoin-superblock_tests.o' failed
make[2]: *** [test/neuralnet/test_test_gridcoin-superblock_tests.o] Error 1
make[2]: Leaving directory '/home/travis/build/gridcoin-community/Gridcoin-Research/build/gridcoin-i686-w64-mingw32/src'
Makefile:8025: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/travis/build/gridcoin-community/Gridcoin-Research/build/gridcoin-i686-w64-mingw32/src'
Makefile:751: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
Build failure. Verbose build follows.

**Comment by cyrossignol** [comment]
@jamescowens Thanks for the log excerpt. I fixed the 32-bit platform issues. 

> As you know, I am a big advocate of powers of 10 seaparation between the precision tiers, because that is how normal "significant digits" work in the scientific word, and they are easy to remember.

I suppose one could argue that it simplifies the protocol to use all the available bits in each size category because we can omit out-of-bounds validation for magnitude tiers. I think the superblock validation against manifest parts covers this for now.

File: src/scraper/scraper.h [hunk 1]
```diff
@@ -70,7 +70,7 @@ int64_t SCRAPER_CMANIFEST_RETENTION_TIME = 48 * 3600;
 bool SCRAPER_CMANIFEST_INCLUDE_NONCURRENT_PROJ_FILES = false;
 double MAG_ROUND = 0.01;
 double NEURALNETWORKMULTIPLIER = 115000;
-double CPID_MAG_LIMIT = 32767;
+double CPID_MAG_LIMIT = NN::Magnitude::MAX;
```
Comment (jamescowens) [inline src/scraper/scraper.h:73]: Yes I agree with this. The default value (absent a protocol override) should be this.

Comment (jamescowens) [reply src/scraper/scraper.h:73]: We should probably put a clamp in ScraperApplyAppCacheEntries() to constrain a mag limit value specified in the protocol to NN:Magnitude::MAX, to ensure we don't break things if someone happens to put a CPID_MAG_LIMIT above the hard coded constant. Thoughts?

**Review by jamescowens — CHANGES_REQUESTED** [review]
Really nice work on this. Only a few minor comments.

**Comment by jamescowens** [comment]
I was wondering whether we should use the same rounding methodology internal to the scraper calcs too, but maybe that is overkill. The current rounding to 0.01 in the scraper internals matches the 100 scale factor for mags less than 1, and it probably will work just as well as it does now.

File: src/scraper/scraper.h (hunk 1 above)
Comment (cyrossignol) [reply src/scraper/scraper.h:73]: I think that's a good idea. If we do, we may want to add reasonable boundaries to some of the other protocol configuration items as well, though.

File: src/neuralnet/superblock.cpp [hunk 2]
```diff
@@ -1197,7 +1226,9 @@ class LegacySuperblockParser
         try {
             m_zero_mags = std::stoi(ExtractXML(packed, "<ZERO>", "</ZERO>"));
         } catch (...) {
-            LogPrintf("LegacySuperblock: Failed to parse zero mag CPIDs");
+            if (fDebug) {
```
[the remark this answers is not in the document]
Comment (cyrossignol) [reply src/neuralnet/superblock.cpp]: Good suggestion--done.

**Comment by cyrossignol** [comment]
Since the scraper data serializes as strings right now, using the same precision tiers will save some bytes. Besides that, I don't see a big reason to change the scraper rounding right now. It will be good to revisit if we change the scraper messages to a binary format.

**Review by jamescowens — APPROVED** [review]

**Review by denravonska — APPROVED** [review]

**Comment by jamescowens** [comment]
Merging.

**Pull request merged**